### PR TITLE
[2.6.5] Add neuvector chart

### PR DESCRIPTION
**Pull request by selvamt94**
* Chart deploys Neuvector components with tag 5.0.0-preview.1 from docker.io registry by default. Images are publicly available, so no need to provide pull secret. Please check for docker rate limit issue if image pull fails.
* Neuvector components are deployed on neuvector-system namespace with default service account. Please create new service account and configure it on values.yaml if neuvector components need to use different service account other than default.
* Set right container runtime from container runtime section. For example, docker runtime for RKE, containerd runtime for RKE2+k8s, k3s for K3s

### CODE REVIEW

**Comment by cbron** [comment]
This shouldn't merge until after 2.6.4 is released. We can get someone from team1 to review if needed.

File: charts/neuvector/1.8.901+up1.8.9/templates/crd.yaml
Comment (cbron) [inline charts/neuvector/1.8.901+up1.8.9/templates/crd.yaml:1]: The CRDs should be split into a secondary chart. It sounds like there is already a CRD-subchart separately, so that can be moved into the second chart. I'll link more info.

File: packages/neuvector/generated-changes/patch/values.yaml.patch
Comment (cbron) [inline packages/neuvector/generated-changes/patch/values.yaml.patch:9]: Need to use system_default_registry: https://github.com/rancher/charts/blob/b55e6ec01907964723942fdbd9f63b9afef7a1ed/packages/rancher-cis-benchmark/charts/templates/_helpers.tpl#L6-L12

File: packages/neuvector/generated-changes/patch/values.yaml.patch
Comment (cbron) [inline packages/neuvector/generated-changes/patch/values.yaml.patch:19]: All of these need to be in format of image: repository/tag
https://github.com/rancher/system-charts/blob/cb84791246fad83dd4bfe909fc7245ea23f69fa1/charts/rancher-monitoring/v0.1.3/charts/prometheus/templates/prometheus.yaml#L26

File: packages/neuvector/generated-changes/patch/values.yaml.patch
Comment (cbron) [inline packages/neuvector/generated-changes/patch/values.yaml.patch:28]: All of these also need to be either in rancher dockerhub or mirrored to rancher dockerhub. I believe we said the latter, so we need to make those repo's and get them mirrored.

File: charts/neuvector/1.8.901+up1.8.9/values.yaml
Comment (cbron) [inline charts/neuvector/1.8.901+up1.8.9/values.yaml:12]: Does this mean its using the default service account ? It needs to use a custom one or it will break CIS benchmarks.

File: charts/neuvector/1.8.901+up1.8.9/Chart.yaml
Comment (cbron) [inline charts/neuvector/1.8.901+up1.8.9/Chart.yaml:19]: We should remove kubeVersion and only use `catalog.cattle.io/kube-version`. 
The reason is because if user has NV installed, and upgrades their chart but not K8s, the standard helm kubeVersion will block edits of the chart, but ours only blocks install.

File: charts/neuvector/1.8.901+up1.8.9/Chart.yaml
Comment (cbron) [inline charts/neuvector/1.8.901+up1.8.9/Chart.yaml:28]: @aiyengar2 should these be 100.0.0+up1.8.9 ? Its a brand new chart, so we could also start at 1.0.0. But I think it should be 100 to be normal.

File: charts/neuvector/1.8.901+up1.8.9/templates/crd.yaml
Comment (cbron) [reply charts/neuvector/1.8.901+up1.8.9/templates/crd.yaml:1]: I believe you can use `generateCRDChart` to do this still. Check other charts package.yaml.

**Review by cbron — CHANGES_REQUESTED** [review]

File: charts/neuvector/1.8.901+up1.8.9/templates/crd.yaml
Comment (garyduan) [reply charts/neuvector/1.8.901+up1.8.9/templates/crd.yaml:1]: There is a separate CRD chart, the purpose of that chart is to allow users to install CRD before installing NeuVector. The core chart has everything, so they can do the installation in one shot.

File: packages/neuvector/generated-changes/patch/values.yaml.patch
Comment (garyduan) [reply packages/neuvector/generated-changes/patch/values.yaml.patch:19]: In the value.yaml, we intentionally make the tag and repo in separate values. This is to prevent users from giving different release versions to controller/enforcer/manager accidentally.

File: packages/neuvector/generated-changes/overlay/app-readme.md
Comment (cbron) [inline packages/neuvector/generated-changes/overlay/app-readme.md:14]: Is this true ? I thought the open-source version worked out of the box. If so, where can they get this from ? 

Also the image pull secret would be invalid because images moved to system_default_registry. Looks like it was removed below.

File: packages/neuvector/generated-changes/overlay/questions.yaml
Comment (cbron) [inline packages/neuvector/generated-changes/overlay/questions.yaml:40]: Preview 3 ?

File: packages/neuvector/generated-changes/overlay/crds/crd.yaml
Comment (cbron) [inline packages/neuvector/generated-changes/overlay/crds/crd.yaml:1]: Is this file actually how the crd chart is being generated ? I was thinking it was because of the crd folder in upstream: https://github.com/neuvector/neuvector-helm/tree/master/charts/crd

**Review by cbron — COMMENTED** [review]
Few more questions then lgtm

File: packages/neuvector/generated-changes/patch/Chart.yaml.patch
Comment (PennyScissors) [inline packages/neuvector/generated-changes/patch/Chart.yaml.patch:13]: this should be `catalog.cattle.io/rancher-version: '>= 2.6.5-0 <= 2.6.100-0'`

File: packages/neuvector/generated-changes/patch/Chart.yaml.patch
Comment (PennyScissors) [inline packages/neuvector/generated-changes/patch/Chart.yaml.patch:11]: this should be `catalog.cattle.io/permits-os: linux,windows` and you will need to make sure the appropriate node-selectors/tolerations are set so the neuvector pods are not scheduled on Windoes nodes.

File: packages/neuvector/generated-changes/patch/Chart.yaml.patch
Comment (PennyScissors) [inline packages/neuvector/generated-changes/patch/Chart.yaml.patch:8]: Does neuvector support k8s 1.22? if it doesn't make sure to set this to `catalog.cattle.io/kube-version: '>=1.18.0-0 < 1.22.0-0`

File: packages/neuvector/generated-changes/overlay/app-readme.md
Comment (selvamt94) [reply packages/neuvector/generated-changes/overlay/app-readme.md:14]: Yes, mistake.  we will remove the pull secret requirement line from the app-readme.

File: packages/neuvector/generated-changes/overlay/questions.yaml
Comment (selvamt94) [reply packages/neuvector/generated-changes/overlay/questions.yaml:40]: Somehow question.yaml is not matching the tag value from values.yaml. The good part is Rancher UI is showing tag value from values.yaml (5.0.0-b1) even though question.yaml has a different value(5.0.0-preview.3).

We will correct it to 5.0.0-b1

File: packages/neuvector/generated-changes/overlay/crds/crd.yaml
Comment (selvamt94) [reply packages/neuvector/generated-changes/overlay/crds/crd.yaml:1]: This crd is the same one that is available in the core chart itself under template. We are using crd.yaml from the core chart for building the feature chart.

This separate crd chart is useful for the customer who wants to deploy CRs from the CRD before deploying core components.

File: packages/neuvector/generated-changes/patch/Chart.yaml.patch
Comment (selvamt94) [reply packages/neuvector/generated-changes/patch/Chart.yaml.patch:13]: we will update it.

File: packages/neuvector/generated-changes/patch/Chart.yaml.patch
Comment (selvamt94) [reply packages/neuvector/generated-changes/patch/Chart.yaml.patch:11]: we agreed with @cbron to use only linux for permit-os now because Neuvector supports only Linux nodes.

File: packages/neuvector/generated-changes/patch/Chart.yaml.patch
Comment (selvamt94) [reply packages/neuvector/generated-changes/patch/Chart.yaml.patch:8]: Neuvector supports 1.22 as well 1.23.

File: packages/neuvector/generated-changes/overlay/app-readme.md
Comment (cbron) [reply packages/neuvector/generated-changes/overlay/app-readme.md:14]: The license key too ?

File: packages/neuvector/generated-changes/overlay/questions.yaml
Comment (cbron) [reply packages/neuvector/generated-changes/overlay/questions.yaml:40]: Default may be overriden by values.yaml, you could just remove it to avoid confusion.

File: packages/neuvector/generated-changes/patch/Chart.yaml.patch
Comment (cbron) [reply packages/neuvector/generated-changes/patch/Chart.yaml.patch:11]: Main question here is how NV reacts to a cluster with Windows nodes, since unlike most apps we have, it needs to interact with those nodes. Since we don't know for sure, I think its best to hold off for now on allowing that use case.

File: charts/neuvector/100.0.0+up2.2.0-b1/values.yaml
Comment (cbron) [reply charts/neuvector/100.0.0+up2.2.0-b1/values.yaml:248]: Yes its fine, as long as it works with images.txt generator.

File: packages/neuvector/generated-changes/overlay/crds/crd.yaml
Comment (cbron) [reply packages/neuvector/generated-changes/overlay/crds/crd.yaml:1]: Ah ok, so it actually comes from the overlay then. Which means every time that file changes you'll have to update the overlay.

File: packages/neuvector/generated-changes/overlay/app-readme.md
Comment (selvamt94) [reply packages/neuvector/generated-changes/overlay/app-readme.md:14]: Yes, license key requirements will also be removed. Shall I update the PR with all these changes now or wait for more reviews from others too?

File: packages/neuvector/generated-changes/patch/Chart.yaml.patch
Comment (PennyScissors) [reply packages/neuvector/generated-changes/patch/Chart.yaml.patch:11]: @cbron We should still add windows to it in order to make neuvector deployable in a hybrid RKE2 cluster. Take for example the CIS benchmark chart which does not support Windows nodes, but you can still deploy it in a hybrid cluster due to these node-selectors/tolerations:

https://github.com/rancher/charts/blob/eaba8696d0e569a64c45ef7f610b501528daae9d/charts/rancher-cis-benchmark/2.0.3/templates/deployment.yaml#L44-L51

https://github.com/rancher/charts/blob/eaba8696d0e569a64c45ef7f610b501528daae9d/charts/rancher-cis-benchmark/2.0.3/templates/_helpers.tpl#L14-L27

File: packages/neuvector/generated-changes/overlay/app-readme.md
Comment (cbron) [reply packages/neuvector/generated-changes/overlay/app-readme.md:14]: Either works, typically people just push updates consistently

Comment (selvamt94) [reply packages/neuvector/generated-changes/overlay/app-readme.md]: I have updated the changes except for this comment https://github.com/rancher/charts/pull/1732#discussion_r848901197

File: packages/neuvector/generated-changes/patch/Chart.yaml.patch
Comment (cbron) [reply packages/neuvector/generated-changes/patch/Chart.yaml.patch:11]: We could try it. My main point above was that those other apps just deal with the linux nodes they get scheduled on. Where as NeuVector might interact with the traffic of all nodes in the hybrid cluster, and thats the real unknown part, how NeuVector reacts in a hybrid cluster at all, regardless of which node in the cluster it gets scheduled on.

**Review by cbron — APPROVED** [review]
lgtm, there is open conversation on mixed windows clusters, but I'd like to address that in follow-up PR when we move 5.0.0-b1 to 5.0.0

**Review by PennyScissors — APPROVED** [review]
Approving since everything else looks good to me, but there is still this question that need to be addressed before merging.
https://github.com/rancher/charts/pull/1732#discussion_r851048894

**Comment by cbron** [comment]
Issue: https://github.com/rancher/rancher/issues/37377

File: packages/neuvector/generated-changes/patch/Chart.yaml.patch
Comment (garyduan) [reply packages/neuvector/generated-changes/patch/Chart.yaml.patch:11]: We do have partial deployment cases, where not all nodes have enforcers running. The network rules are learned/defined and enforced accordingly. Having said that, the network function works the best in full deployment. I suggest we disable neuvector in mixed environment for now, and update the chart later.

**Pull request merged**
